derrida/interventions/management/commands/insertion_data.py

```diff
@@ -0,0 +1,145 @@
+            group_label = '%s. %s' % (canvas.manifest.label.strip('.'), insertion_match.group('label'))
+            insertion_images[group_label].append(canvas)
+
+        data = [self.insertion_data(label, canvas_group)
+                for label, canvas_group in insertion_images.items()]
+
+        # filter out any null values for canvases not linked to work instance
+        data = [d for d in data if d]
+
+        # list of dictionaries can be output as is for JSON export
+        with open('{}.json'.format(self.base_filename), 'w') as jsonfile:
+            json.dump(data, jsonfile, indent=2)
+
+        # generate CSV export
+        with open('{}.csv'.format(self.base_filename), 'w') as csvfile:
+            # write utf-8 byte order mark at the beginning of the file
+            csvfile.write(codecs.BOM_UTF8.decode())
+
+            csvwriter = csv.DictWriter(csvfile, fieldnames=self.csv_fields)
+            csvwriter.writeheader()
+
+            for insertion in data:
+                csvwriter.writerow(self.flatten_dict(insertion))
+
+    def insertion_data(self, label, canvases):
+        '''Generate a dictionary of data to export for a group
+        of canvases representing a single insertion in a single book.'''
+
+        # NOTE: using OrderedDict to ensure JSON output follows logical
+        # order in Python < 3.6, where dict order is not guaranteed
+
+        first_canvas = canvases[0]
+        # access book via manifest instance (reverse of digital edition relation)
+        # at least one insertion canvas is not related to a book...
+        try:
+            book = first_canvas.manifest.instance
+        except ObjectDoesNotExist:
+            print('manifest %s not related to work instance' % first_canvas.manifest.label)
+            # NOTE: one manifest is not linked, and it seems to be a duplicate;
+            # omit from export
+            book = None
+            return
+
+        # extract page number from first canvas label
+        page = RE_INSERTION_LABEL.match(first_canvas.label).group('page')
+
+        return OrderedDict([
+            ('id', label),   # provisional
+            ('intervention_type', 'insertion'),
+            ('book', OrderedDict([
+                ('id', book.get_uri()),
+                ('title', book.display_title()),
+                ('type', book.item_type)
+            ])),
+            ('page', page.strip()),
+            ('num_images', len(canvases)),
+            # to avoid repetition, only include unique portion of the label
+            # (i.e., recto/verso or roman numerals for multipage items)
+            ('image_labels', [RE_INSERTION_LABEL.match(c.label).group('insertion_page').strip()
+                              for c in canvases]),
+            # ('image_labels', [c.label for c in canvases]),
+            # TODO: use proxied iiif image url
+            # FIXME: iiif image urls currently not resolving
+            ('image_iiif', [str(c.image) for c in canvases])
```

Comment:
This you'll be able to copy and paste from my PR when I'm done, right?

Reply:
yes, that was my plan! 😁